bertopic/_bertopic.py

```diff
@@ -439,14 +466,23 @@ def fit_transform(self,
             # Save the top 3 most representative documents per topic
             self._save_representative_docs(documents)
 
+        # In the case of zero-shot topics, probability will come from cosine similarity,
+        # and the HDBSCAN model will be removed
+        if self._is_zeroshot() and len(assigned_documents) > 0:
+            self.hdbscan_model = BaseCluster()
+            sim_matrix = cosine_similarity(embeddings, np.array(self.topic_embeddings_))
+
+            if self.calculate_probabilities:
+                probabilities = sim_matrix
+            else:
+                # Use `topics_before_reduction` because `self.topics_` may have already been updated from
+                # reducing topics, and the original probabilities are needed for `self._map_probabilities()`
+                probabilities = sim_matrix[np.arange(len(documents)), topics_before_reduction]
```

Comment:
Could this be simplified by moving it to the end of the function and running something like:

```python
if self._is_zeroshot() and len(assigned_documents) > 0:
            self.hdbscan_model = BaseCluster()
            _, probabilities = self.transform(assigned_documents, embeddings)
```

That way, it would actually be a more proper `.fit` + `.transform` combo as you typically see in scikit-learn and reduces the lines of code a bit.

Reply:
This goes back to our discussion on using cosine similarities for probabilities after zero-shot topic modeling.

> They are indeed not perfectly aligned with one another. With flat probabilities/similarities, this is not an issue as we can simply give back the similarity of the assigned topic. With topic-doc probabilities/similarities, I'm not sure how to solve that issue without attempting to create a separate model (i.e., logistic regression) that is trained (and perhaps even overfitted) on the embeddings (X) and assigned topic (y) to extract hopefully more accurate probabilities.

If using `.transform`, the assignments, and the corresponding probabilities will not be aligned with the current state of assignments, because the current state came from a mix of zero-shot and clustering, and `self.topic_embeddings_` was not used at all when making these assignments.

We agreed this misalignment is fine. We will return the cosine similarities for the current assignments, even if calculating similarities using `self.topic_embeddings_` would show different assignments. I am not sure it is possible to reuse `.transform` for this purpose. I'm open to suggestions, but having that bit of duplicated code seems like the simplest solution.

On a related note, I do not think my solution accounts for outliers.
I have
```
probabilities = sim_matrix[
                    np.arange(len(documents)), topics_before_reduction
                ]
```
but I think it should be
```
probabilities = sim_matrix[
                    np.arange(len(documents)), np.array(topics_before_reduction) + self._outliers
                ]
```
Does this make sense to you?

Reply:
> From a user perspective, I would expect the assignments from .fit_transform to be identical to the assignments from .fit followed by .transform.

Agreed, that wasn't the case previously (also due to some specific behavior of HDBSCAN) but that would definitely be preferred. 

> I'm open to suggestions, but having that bit of duplicated code seems like the simplest solution. 

Ah right, I totally missed that! Yep, let's keep it as is. 

> On a related note, I do not think my solution accounts for outliers.

Yeah, that should indeed be `+ self._outliers` I believe since the lowest potential value is -1 there which does nasty things when indexing.